internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,211 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/contract"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+// preflightFailedRequeueAfter is used to requeue the MachineSet to re-verify the preflight checks if
+// the preflight checks fail.
+const preflightFailedRequeueAfter = 15 * time.Second
+
+func (r *Reconciler) runPreflightChecks(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	// If the MachineSetPreflightChecks feature gate is disabled return early.
+	if !feature.Gates.Enabled(feature.MachineSetPreflightChecks) {
+		return ctrl.Result{}, nil
+	}
+
+	skipped := skippedPreflightChecks(ms)
+	// If all the preflight checks are skipped then return early.
+	if skipped.Has(clusterv1.MachineSetPreflightCheckAll) {
+		return ctrl.Result{}, nil
+	}
+
+	// If the cluster does not have a control plane reference then there is nothing to do. Return early.
+	if cluster.Spec.ControlPlaneRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the control plane object.
+	controlPlane, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to get ControlPlane %s", klog.KRef(cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name))
+	}
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// If the Control Plane version is not set then we are dealing with a control plane that does not support version
+	// or a control plane where the version is not set. In both cases we cannot perform any preflight checks as
+	// we do not have enough information. Return early.
+	cpVersion, err := contract.ControlPlane().Version().Get(controlPlane)
+	if err != nil {
+		if errors.Is(err, contract.ErrFieldNotFound) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to get the version of ControlPlane %s", cpKlogRef)
+	}
+	cpSemver, err := semver.ParseTolerant(*cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to parse ControlPlane version %s", *cpVersion)
```

Comment:
```suggestion
		return ctrl.Result{}, errors.Wrapf(err, "failed to perform preflight checks: failed to parse version %q of ControlPlane %s", *cpVersion, cpKlogRef)
```

Reply:
Probably the same for the MachineSet error in l.101. It's a bit redundant in this case as the logger will have the MachineSet as well, but it would be consistent with this error here